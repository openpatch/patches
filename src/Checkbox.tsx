// follows https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-1/checkbox-1.html
import { KeyboardEvent, ReactNode } from "react";
import { css } from "@emotion/react";
import { Box } from "./Box";

export type CheckboxProps = {
  id?: string;
  children?: ReactNode;
  checked?: boolean;
  disabled?: boolean;
  required?: boolean;
  onChange?: (checked: boolean) => void;
  onBlur?: () => void;
};

export const Checkbox = ({
  id,
  children,
  checked = false,
  disabled = false,
  required = false,
  onChange = () => {},
  onBlur = () => {},
}: CheckboxProps) => {
  function handleClick() {
    if (!disabled) {
      onChange(!checked);
    }
  }

  function handleKeyDown(e: KeyboardEvent<HTMLDivElement>) {
    if (!disabled) {
      if (e.key === " ") {
        onChange(!checked);
        e.stopPropagation();
        e.preventDefault();
      }
    }
  }

  return (
    <Box
      id={id}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      onBlur={onBlur}
      css={(theme) => [
        css`
          display: inline-block;
          position: relative;
          padding-left: 1.4em;
          cursor: default;

          ::before,
          ::after {
            position: absolute;
            top: 50%;
            left: 7px;
            transform: translate(-50%, -50%);
            content: "";
          }

          ::before {
            width: 18px;
            height: 18px;
            border: 1px solid hsl(0, 0%, 66%);
            border-radius: 0.2em;
          }

          :active::before {
            background-image: linear-gradient(
              to bottom,
              hsl(300, 3%, 73%),
              hsl(300, 3%, 93%) 30%
            );
          }

          :focus {
            outline: none;
          }

          :focus::before {
            width: 20px;
            height: 20px;
            box-sizing: content-box;
            box-shadow: ${theme.shadows.outline};
          }
        `,
        checked &&
          css`
            ::before {
              background: ${theme.colors.primary["800"]};
            }

            ::after {
              display: block;
              width: 0.25em;
              height: 0.5em;
              border: solid #fff;
              border-width: 0 0.155em 0.155em 0;
              transform: translateY(-65%) translateX(-50%) rotate(45deg);
            }

            :active::before {
              background: ${theme.colors.primary["800"]};
            }
          `,
        disabled &&
          css`
            cursor: not-allowed;

            ::before {
              background: ${theme.colors.neutral["100"]};
            }
            :focus::before {
              box-shadow: none;
            }

            :active::before {
              background-image: none;
            }
          `,
      ]}
      role="checkbox"
      aria-checked={checked}
      aria-disabled={disabled}
      aria-required={required}
      tabIndex={0}
    >
      {children}
    </Box>
  );
};
