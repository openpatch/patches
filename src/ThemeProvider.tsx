import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
