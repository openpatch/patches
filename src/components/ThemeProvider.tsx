import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
