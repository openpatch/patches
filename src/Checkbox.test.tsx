
import { render } from "@testing-library/react";
import { Checkbox, CheckboxProps } from "./Checkbox";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: CheckboxProps) =>
render(
    <ThemeProvider>
        <Checkbox {...props} />
    </ThemeProvider>
);

test("should render", () => {
   renderWithTheme();
});

test("should match snapshot", () => {
    const { container } = renderWithTheme();
    expect(container).toMatchSnapshot();
});