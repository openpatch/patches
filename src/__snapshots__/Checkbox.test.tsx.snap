// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  display: inline-block;
  position: relative;
  padding-left: 1.4em;
  cursor: default;
}

.emotion-0::before,
.emotion-0::after {
  position: absolute;
  top: 50%;
  left: 7px;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  content: "";
}

.emotion-0::before {
  width: 18px;
  height: 18px;
  border: 1px solid hsl(0, 0%, 66%);
  border-radius: 0.2em;
}

.emotion-0:active::before {
  background-image: linear-gradient(
                to bottom,
                hsl(300, 3%, 73%),
                hsl(300, 3%, 93%) 30%
              );
}

.emotion-0:focus {
  outline: none;
}

.emotion-0:focus::before {
  width: 20px;
  height: 20px;
  box-sizing: content-box;
  box-shadow: 0 0 0 3px rgba(66, 153, 225, 0.5);
}

<div>
  <div
    id="patches-root"
  >
    <div
      aria-checked="false"
      aria-disabled="false"
      aria-required="false"
      class="emotion-0"
      role="checkbox"
      tabindex="0"
    />
  </div>
</div>
`;
